File: 프로그래머스/2/17677. ［1차］ 뉴스 클러스터링/［1차］ 뉴스 클러스터링.py
from collections import defaultdict
def solution(str1, str2):
    answer = 0
    dict_one=defaultdict(int)
    dict_two=defaultdict(int)
    for i in range(len(str1)-1):
        string=(str1[i]+str1[i+1]).lower()
        if string.isalpha():
            dict_one[string]+=1
    for i in range(len(str2)-1):
        string=(str2[i]+str2[i+1]).lower()
        if (string).isalpha():
            dict_two[string]+=1
    
    set_one=set(dict_one.keys())
    set_two=set(dict_two.keys())
    
    inter=set_one.intersection(set_two)
    un=set_one.union(set_two)
    
    up=0
    down=0
    for s in inter:
        up+=min(dict_one[s],dict_two[s])
    for s in un:
        down+=max(dict_one[s],dict_two[s])
  
    return int((up/down)*65536) if down!=0 else 65536

File: 프로그래머스/2/17677. ［1차］ 뉴스 클러스터링/［1차］ 뉴스 클러스터링.java
// 개수 기준, 교집합은 둘 다 있는거 중에 작은값 / 합집합은 다 넣고 둘다 있다면 큰값 (근데 그냥 요소 다 더하고 교집합 개수 빼면 됨!)
import java.util.*;
class Solution {
    public int solution(String str1, String str2) {
        Map<String,Integer> countsOne=countElements(str1);
        Map<String,Integer> countsTwo=countElements(str2);
        int intersection = getIntersection(countsOne, countsTwo);
        int union = getUnion(countsOne, countsTwo)-intersection;
        
        if (union==0){
            return 65536;
        }
        
        return (int)(((double)intersection/union)*65536);
    }
    
    private Map<String,Integer> countElements(String str){
        Map<String,Integer> counts = new HashMap<>();
        for (int i=0;i<str.length()-1;i++){
            String element = str.substring(i,i+2).toLowerCase();
            if (isString(element)){
                counts.put(element,counts.getOrDefault(element,0)+1);
            }
        }
        return counts;
    }
    
    private boolean isString(String s){
        return Character.isLetter(s.charAt(0)) && Character.isLetter(s.charAt(1));
    }
    
    private int getIntersection(Map<String,Integer> countsOne, Map<String,Integer> countsTwo){
        int total = 0;
        for(String key : countsOne.keySet()){
            if (countsTwo.containsKey(key)){
                total+=Math.min(countsOne.get(key),countsTwo.get(key));
            }
        }
        return total;
    }
    
    private int getUnion(Map<String,Integer> countsOne, Map<String,Integer> countsTwo){
        int total = 0;
        for (Integer i: countsOne.values()){
            total+=i;
        }
        for (Integer i: countsTwo.values()){
            total+=i;
        }
        return total;
    }
    
}